File: localization/budget.go
// Package localization (TTTT-I): per-request localization budget +
// tracing. Counts the number of cache-miss override lookups, cache
// reads, and provider calls a single request issues, then surfaces
// the totals so observability dashboards + a hard-fail dev gate
// can catch N+1 regressions before they hit prod.
//
// Wire-up: Seam.Translate + Seam.TranslateBatch increment the per-
// kind counter on the ctx; an Echo middleware reads the counts after
// the handler returns and stamps them as response headers. Optional
// dev-mode hard fail via WithBudget(ctx, limits) returns an error
// when the limits are exceeded — the handler can then either log or
// fail the request loudly.

package localization

import (
	"context"
	"sync/atomic"
)

// BudgetKind enumerates the cost categories the seam tracks.
type BudgetKind int

const (
	// BudgetKindOverrideLookup counts override DB queries (post-TTTT-B:
	// 1 composite SELECT per batch call; pre-TTTT-B: 1 per ref).
	BudgetKindOverrideLookup BudgetKind = iota
	// BudgetKindCacheGet counts Redis GET/MGET round-trips.
	BudgetKindCacheGet
	// BudgetKindProviderCall counts LLM / opus-mt provider calls
	// (per Sweep TTTT-B: typically 1 per batch).
	BudgetKindProviderCall
	// BudgetKindRefResolved counts total refs resolved (regardless of
	// path). A single batch call resolves N refs but only issues
	// O(1) round-trips, so this metric is the most useful for
	// detecting "fan-out got worse" regressions.
	BudgetKindRefResolved
	budgetKindCount
)

// budgetCounters is the per-request mutable counter. Stored on ctx
// behind a private key; readers use BudgetSnapshot to read.
type budgetCounters struct {
	values [budgetKindCount]int64
}

func (b *budgetCounters) add(kind BudgetKind, n int) {
	if b == nil || kind >= budgetKindCount {
		return
	}
	atomic.AddInt64(&b.values[kind], int64(n))
}

func (b *budgetCounters) snapshot() BudgetSnapshot {
	if b == nil {
		return BudgetSnapshot{}
	}
	return BudgetSnapshot{
		OverrideLookups: int(atomic.LoadInt64(&b.values[BudgetKindOverrideLookup])),
		CacheGets:       int(atomic.LoadInt64(&b.values[BudgetKindCacheGet])),
		ProviderCalls:   int(atomic.LoadInt64(&b.values[BudgetKindProviderCall])),
		RefsResolved:    int(atomic.LoadInt64(&b.values[BudgetKindRefResolved])),
	}
}

// BudgetSnapshot is the read-side view of the per-request counters.
// Cheap to copy; safe to embed in response payloads / logs.
type BudgetSnapshot struct {
	OverrideLookups int `json:"override_lookups"`
	CacheGets       int `json:"cache_gets"`
	ProviderCalls   int `json:"provider_calls"`
	RefsResolved    int `json:"refs_resolved"`
}

type budgetCtxKey struct{}

// WithBudget attaches a per-request counter to ctx. Returns a new ctx;
// callers should propagate this to every downstream localization call.
//
// The intended pattern is: an Echo middleware calls WithBudget on
// every request, then reads BudgetFromContext after the handler returns
// and stamps the totals as X-Localization-* response headers (visible
// to mobile dev tools + observability scrapers).
func WithBudget(ctx context.Context) context.Context {
	if BudgetFromContext(ctx) != nil {
		return ctx // already wired; don't reset
	}
	return context.WithValue(ctx, budgetCtxKey{}, &budgetCounters{})
}

// BudgetFromContext returns the active counters, or nil if WithBudget
// was never called on this ctx (e.g. background jobs, internal tests).
func BudgetFromContext(ctx context.Context) *budgetCounters {
	if ctx == nil {
		return nil
	}
	bc, _ := ctx.Value(budgetCtxKey{}).(*budgetCounters)
	return bc
}

// BudgetSnapshotFromContext returns the current snapshot, or a zero
// snapshot if no budget is wired.
func BudgetSnapshotFromContext(ctx context.Context) BudgetSnapshot {
	bc := BudgetFromContext(ctx)
	if bc == nil {
		return BudgetSnapshot{}
	}
	return bc.snapshot()
}

// RecordBudget bumps the counter for `kind` by `n`. No-op when no
// budget is wired (background workers). The seam calls this internally
// from Translate / TranslateBatch / the batch phase helpers; external
// code shouldn't need to.
func RecordBudget(ctx context.Context, kind BudgetKind, n int) {
	bc := BudgetFromContext(ctx)
	if bc == nil {
		return
	}
	bc.add(kind, n)
}

File: kielo_shared/localization/budget.py
"""Per-request localization budget — Python sibling of Go's
``kielo-shared/localization/budget.go``.

Sweep YYYY: mirrors the TTTT-I budget infrastructure on the Python
side so the engine's localization seam ALSO reports per-request fan-
out costs via the same ``X-Kielo-Loc-*`` response headers. Pre-YYYY
the engine was invisible — only kielo-convo (Go) emitted budget
headers, so dashboards monitoring scenarios couldn't compare
against the engine's roadmap / curriculum / discovery hot paths.

Wire-up:

1. ``LocalizationBudgetMiddleware`` opens a per-request counter on
   the contextvar, runs the handler, reads the snapshot, stamps
   ``X-Kielo-Loc-Refs/Overrides/CacheGets/Providers`` headers.

2. Hot paths call ``record_budget(kind, n)`` at each phase. Today
   the engine's ``content_localizer.localize_batch`` is the
   single seam entry; later sweeps can instrument the structured
   content localizer batch helpers + reusable field localizer to
   distinguish "1 ref via 1 LLM call" from "N refs via 1 LLM call".

3. Background tasks (Pub/Sub workers, cron) don't wire the
   middleware, so ``record_budget`` becomes a no-op for them.
   Same shape as the Go side's behavior under WithBudget-not-set.
"""

from __future__ import annotations

from contextvars import ContextVar, Token
from dataclasses import dataclass
from enum import IntEnum
from typing import Optional


class BudgetKind(IntEnum):
    """Categories the localization seam tracks per-request."""

    OVERRIDE_LOOKUP = 0  # override-store DB queries
    CACHE_GET = 1  # Redis GET/MGET round-trips
    PROVIDER_CALL = 2  # LLM / opus-mt provider calls
    REF_RESOLVED = 3  # total refs resolved (a single batch call resolves N refs but issues O(1) RTTs)


@dataclass
class _BudgetCounters:
    """Mutable per-request counters. No atomic needed under CPython's
    GIL for the integer-add pattern; ContextVar isolation handles the
    cross-task case (each asyncio task sees its own ctx)."""

    override_lookups: int = 0
    cache_gets: int = 0
    provider_calls: int = 0
    refs_resolved: int = 0

    def add(self, kind: BudgetKind, n: int) -> None:
        if kind == BudgetKind.OVERRIDE_LOOKUP:
            self.override_lookups += n
        elif kind == BudgetKind.CACHE_GET:
            self.cache_gets += n
        elif kind == BudgetKind.PROVIDER_CALL:
            self.provider_calls += n
        elif kind == BudgetKind.REF_RESOLVED:
            self.refs_resolved += n

    def snapshot(self) -> "BudgetSnapshot":
        return BudgetSnapshot(
            override_lookups=self.override_lookups,
            cache_gets=self.cache_gets,
            provider_calls=self.provider_calls,
            refs_resolved=self.refs_resolved,
        )


@dataclass(frozen=True)
class BudgetSnapshot:
    """Read-side view of the per-request counters.

    Mirrors the Go ``BudgetSnapshot`` struct field-for-field so
    downstream observability code that aggregates across services
    can use the same shape.
    """

    override_lookups: int = 0
    cache_gets: int = 0
    provider_calls: int = 0
    refs_resolved: int = 0


_active_budget: ContextVar[Optional[_BudgetCounters]] = ContextVar(
    "kielo_localization_budget", default=None
)


def with_budget() -> Token:
    """Attach a fresh per-request counter to the current context.

    Mirrors ``WithBudget(ctx)`` in Go. Returns a ContextVar Token the
    caller MUST pass to ``reset_budget`` in a try/finally (or use the
    LocalizationBudgetMiddleware which handles it automatically).

    Idempotent: if a budget is already wired on this ctx, returns a
    Token that resets to the existing value (so calling with_budget
    inside a nested scope doesn't double-stamp).
    """
    existing = _active_budget.get()
    if existing is not None:
        # Already wired — reuse counters; reset will restore current
        # value (which is itself, so it's a no-op token).
        return _active_budget.set(existing)
    return _active_budget.set(_BudgetCounters())


def reset_budget(token: Token) -> None:
    """Restore the budget contextvar to its pre-with_budget state."""
    _active_budget.reset(token)


def budget_from_context() -> Optional[_BudgetCounters]:
    """Return the active counters, or None if no budget is wired."""
    return _active_budget.get()


def budget_snapshot() -> BudgetSnapshot:
    """Return the current snapshot, or a zero snapshot if no budget."""
    counters = _active_budget.get()
    if counters is None:
        return BudgetSnapshot()
    return counters.snapshot()


def record_budget(kind: BudgetKind, n: int = 1) -> None:
    """Increment the counter for ``kind`` by ``n``. No-op when no
    budget is wired (background workers, scripts, unit tests).

    Same shape as Go's ``RecordBudget(ctx, kind, n)``. Called by the
    seam / batch helpers at each phase; external callers shouldn't
    need to.
    """
    counters = _active_budget.get()
    if counters is None:
        return
    counters.add(kind, n)
